pelican/plugins/seo/tests/test_seo_enhancer.py

```diff
@@ -123,6 +123,188 @@ def test_add_html_enhancements_to_file(self, fake_article, fake_seo_enhancer):
 </html>"""
             )
 
+    def test_add_html_enhancements_to_file_with_save_as(
```

Comment:
You should test first the behaviour of SEO enhancer in case of external canonical (`save_as` is already tested [here](https://github.com/pelican-plugins/seo/blob/master/pelican/plugins/seo/tests/test_canonical_url_creator.py#L19). You can duplicate this test to test `external_canonical`.

About HTML generation, indeed canonical feature is not tested yet. But I think it's better to parametrize a test to avoid duplication, something like:
```
@pytest.mark.parametrize("type,value", [
    (external_canonical,link)
    (save_as,file_url),
    both,
    nothing_settled,
])
def test_add_html_enhancements_to_file_canonical_link(self, type, value):
    fake_article["metadata"][type] = value
```
Just a draft.

Reply:
I will study this when I have more time later.

Reply:
I'll create an issue too for this topic, if someone want to handle it one day.

Reply:
#35